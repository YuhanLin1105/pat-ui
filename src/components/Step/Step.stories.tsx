import React from 'react';
import {action} from '@storybook/addon-actions';
// import StepItem from './StepItem';
import Step from './Step'
import {steps} from './data.mock'

export default {
    title: 'Steps',
    component: Step,
  };
  
 
export const horizontalSteps = ()=>{
    return(
    <div className="App" onClick={action('StepItem')}>
    <Step stepStyle='horizontal' 
    step= { steps } >Horizontal steps:</Step> 
  </div>
    )
}
export const verticalSteps = ()=>{
    return(
    <div className="App" onClick={action('StepItem')}>
    <Step stepStyle='vertical' 
     step= { steps } >Vertical Steps:</Step>
  </div>
    )
}
export const SizeVariations = ()=>{
  return(
  <div className="App" onClick={action('StepItem')}>
  <div >
    <Step stepStyle='horizontal' stepSize='sm'  step= { steps } >Small Size:</Step>
  </div>
  <br/>
  <br/>
  <div >
    <Step stepStyle='horizontal' stepSize='md'  step= { steps } >Medium Size:</Step>
  </div>
  <br/>
  <div >
    <Step stepStyle='horizontal' stepSize='lg'  step= { steps } >Large Size:</Step>
  </div>
  <br/>
  </div>
  
  )
}