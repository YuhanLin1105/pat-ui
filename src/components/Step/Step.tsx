import React, { FC } from 'react';
import { classNames } from '../../utils/classNames'
import Icon from '../Icon/Icon'

export type StepperSize = 'lg' | 'md' | 'sm';
interface IStepper {
  stepperType?: 'horizontal' | 'vertical';
  stepperSize?: StepperSize;
}

/**
 * A Stepper consists of multiple steps and indicates a series of activities.
 *
 * ```js
 * import { Stepper, Step } from 'pat-ui'
 * ```
 */
export const Stepper: FC<IStepper> = (props) => {
  const { children, stepperType, stepperSize } = props

  // add order to Steps
  const childrenWithProps = React.Children.map(children, (child, index) => {
    // checking isValidElement is the safe way and avoids a typescript error too
    if (React.isValidElement(child)) {
      return React.cloneElement(child, { order: index + 1 });
    }
    return child;
  });

  // add connections between steps
  const content = React.Children.toArray(childrenWithProps).flatMap(
    (value, index, array) =>
      array.length - 1 !== index // check for the last item
        ? [value, <StepConnection key={index} connectionType={stepperType}></StepConnection>]
        : value,
  );
  return (
    <div className={`stepper stepper-${stepperType} stepper-${stepperSize}`} data-testid='stepper-element'>{content}</div>
  )
}

Stepper.defaultProps = {
  stepperType: 'horizontal',
  stepperSize: 'md',
};

export interface IStepConnection {
  connectionType?: 'horizontal' | 'vertical';
}

export const StepConnection: FC<IStepConnection> = (props) => {
  const { connectionType } = props
  let content;
  if (connectionType === 'horizontal') {
    content = <hr />
  }
  else if (connectionType === 'vertical') {
    content = <div className='vl'></div>
  }
  return (
    <div className='step-connection'>
      {content}
    </div>
  )
}

export type StepStatus =
  | 'active'
  | 'completed'
  | 'incompleted';

export interface IStep {
  order?: number | string;
  stepStatus?: StepStatus;
}

export const Step: FC<IStep> = (props) => {
  const { children, order, stepStatus } = props;

  // integrate props to string
  let styleClasses = classNames('step', {
    [`step-${stepStatus}`]: true,
    // [`btn-${btnSize}`]: !!btnSize // undefined -> false 
  });
  const svgContent = stepStatus === 'completed' ?
    <Icon
      className='step-icon-check'
      disabled
      loading={false}
      name='check'
    /> :
    <>
      <circle cx='12' cy='12' r='12' className='step-icon-circle'></circle>
      <text className='step-icon-text' x='12' y='17' textAnchor='middle'>
        {order}
      </text>
    </>

  return (
    <div className={styleClasses} data-testid={`step-${order}-element`}>
      <span className='step-icon-container'>
        <svg className='step-icon' height='28' viewBox='0 0 24 24'>

          {svgContent}
        </svg>
      </span>
      <span className='step-label-container'>
        <span className='step-label'>{children}</span>
      </span>
    </div>
  )
}

Step.defaultProps = {
  stepStatus: 'incompleted',
};

