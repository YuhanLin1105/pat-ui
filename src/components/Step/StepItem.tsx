import React, { FC } from 'react';

interface Isteps {
  label?: string;
  description?: string;
  icons?: any;
  className?: boolean;
  handleClick?: Function;
}

const StepItem: FC<Isteps> = (props) => {
  const { label, description, icons } = props;
  return (
    <div className='step-item'>
        <div className="step-icon">
            {icons}
        </div>
        <div className='step-text'>
            <h6><b>{label}</b></h6>
            <p><i>{description}</i></p>
        </div>
    </div>
    // <li className="active">{label}</li>
  );
};

export default StepItem;
