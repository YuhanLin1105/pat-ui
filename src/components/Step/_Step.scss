.steps-wrapper {
  width: 80%;

  margin: 0 auto;
}

.stepper { 
  display: flex;
  padding: 1.5rem;
  border: 1px solid;
  border-color: #ced4da;
}

.stepper-horizontal {
  align-items: center;
  flex-direction: row;
}
.stepper-vertical { 
  flex-direction: column;
}


// step
.step {
  display: flex;
  align-items: center; 
  padding: 1rem;
}

// step icon
.step-icon-container {
  display: flex;
  align-items: center;
}

.step-icon {
  padding-right: 1rem;
}

.step-incompleted .step-icon-circle {
  fill:#ced4da
}

.step-active .step-icon-circle {
  fill:turquoise
}

.step-icon-text {
  fill: white;
}


// step-connection
.step-connection {
  flex: 1 1 auto;
}

.step-connection>.vl {
  width: 0;
  margin-left: 1.8rem;
  border-left: 2px solid #ced4da;
  border-right: 2px solid #ced4da;
  border-radius: 2px;
  min-height: 50px;
}

.step-connection>hr {
  border: 2px solid #ced4da;
  border-radius: 2px;
}
