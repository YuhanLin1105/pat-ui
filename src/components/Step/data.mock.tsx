import Icon from '../Icon';
import React from 'react';

 export const steps =  [ {
    id: 0,
    label: 'Step 1',
    description: 'Steppers display progress',
    icon: <Icon name='info'/>,
    
},
{
    id:1,
    label: 'Step 2',
    description: 'Steppers display progress',
    icon: <Icon name='info'/>,
    
},
{
    id:2,
    label: 'Step 3',
    description: 'Steppers display progress',
    icon: <Icon name='info'/>,
    
}
] 