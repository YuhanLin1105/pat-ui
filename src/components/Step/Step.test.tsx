import React from 'react'
import { render } from '@testing-library/react'
import { Stepper, Step } from './Step'

// shallow test 
describe('Stepper', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(
      <Stepper>
        <Step stepStatus='active'>Step1</Step>
        <Step>Step2</Step>
        <Step>Step3</Step>
      </Stepper>
    ); // like a DOM
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render horizontal stepper with active and incompleted steps', () => {
    const wrapper = render(
      <Stepper>
        <Step stepStatus='active'>Step1</Step>
        <Step>Step2</Step>
        <Step>Step3</Step>
      </Stepper>
    ); // like a DOM
    
    // test Stepper
    const stepperElement = wrapper.getByTestId('stepper-element') as HTMLElement;
    expect(stepperElement).toBeInTheDocument();
    expect(stepperElement).toHaveClass('stepper stepper-horizontal')
    
    // test Step
    const activeStepElement  = wrapper.getByText('1').closest('div') as HTMLElement;
    expect(activeStepElement).toHaveClass('step step-active')
    const incompletedStepElement  = wrapper.getByText('2').closest('div') as HTMLElement;
    expect(incompletedStepElement).toHaveClass('step step-incompleted')
    
    const stepLabelElement = wrapper.queryByText('Step1') as HTMLElement;
    expect(stepLabelElement ).toBeInTheDocument();
    expect(stepLabelElement.tagName).toBe('SPAN')
    expect(stepLabelElement ).toHaveClass('step-label')
  });

  it('should render vertical stepper with active and incompleted steps', () => {
    const wrapper = render(
      <Stepper stepperType='vertical'>
        <Step stepStatus='completed'>Step1</Step>
        <Step stepStatus='active'>Step2</Step>
        <Step>Step3</Step>
      </Stepper>
    ); // like a DOM
    
    const stepperElement = wrapper.getByTestId('stepper-element') as HTMLElement;
    expect(stepperElement).toHaveClass('stepper stepper-vertical')
    const completedStepElement  = wrapper.getByTestId('step-1-element') as HTMLElement;
    expect(completedStepElement).toHaveClass('step step-completed')
    const activeStepElement  = wrapper.getByText('2').closest('div') as HTMLElement;
    expect(activeStepElement).toHaveClass('step step-active')
    const incompletedStepElement  = wrapper.getByText('3').closest('div') as HTMLElement;
    expect(incompletedStepElement).toHaveClass('step step-incompleted')
    
    const stepLabelElement = wrapper.queryByText('Step1') as HTMLElement;
    expect(stepLabelElement ).toBeInTheDocument();
    expect(stepLabelElement.tagName).toBe('SPAN')
    expect(stepLabelElement ).toHaveClass('step-label')
  });
})