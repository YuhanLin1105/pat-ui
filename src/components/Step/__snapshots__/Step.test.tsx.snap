// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="stepper stepper-horizontal stepper-md"
    data-testid="stepper-element"
  >
    <div
      class="step step-active"
      data-testid="step-1-element"
    >
      <span
        class="step-icon-container"
      >
        <svg
          class="step-icon"
          height="28"
          viewBox="0 0 24 24"
        >
          <circle
            class="step-icon-circle"
            cx="12"
            cy="12"
            r="12"
          />
          <text
            class="step-icon-text"
            text-anchor="middle"
            x="12"
            y="17"
          >
            1
          </text>
        </svg>
      </span>
      <span
        class="step-label-container"
      >
        <span
          class="step-label"
        >
          Step1
        </span>
      </span>
    </div>
    <div
      class="step-connection"
    >
      <hr />
    </div>
    <div
      class="step step-incompleted"
      data-testid="step-2-element"
    >
      <span
        class="step-icon-container"
      >
        <svg
          class="step-icon"
          height="28"
          viewBox="0 0 24 24"
        >
          <circle
            class="step-icon-circle"
            cx="12"
            cy="12"
            r="12"
          />
          <text
            class="step-icon-text"
            text-anchor="middle"
            x="12"
            y="17"
          >
            2
          </text>
        </svg>
      </span>
      <span
        class="step-label-container"
      >
        <span
          class="step-label"
        >
          Step2
        </span>
      </span>
    </div>
    <div
      class="step-connection"
    >
      <hr />
    </div>
    <div
      class="step step-incompleted"
      data-testid="step-3-element"
    >
      <span
        class="step-icon-container"
      >
        <svg
          class="step-icon"
          height="28"
          viewBox="0 0 24 24"
        >
          <circle
            class="step-icon-circle"
            cx="12"
            cy="12"
            r="12"
          />
          <text
            class="step-icon-text"
            text-anchor="middle"
            x="12"
            y="17"
          >
            3
          </text>
        </svg>
      </span>
      <span
        class="step-label-container"
      >
        <span
          class="step-label"
        >
          Step3
        </span>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
